payments_service/auth/forms.py

```diff
@@ -13,32 +13,76 @@
-        log.info(u'verifying access token at {url}; token={token}'
-                 .format(url=url, token=access_token))
+    def clean(self):
+        data = self.cleaned_data
+        if not (data.get('access_token') or data.get('authorization_code')):
+            raise forms.ValidationError(
+                'access_token or authorization_code is required'
+            )
+        return data
+
+    def fxa_post(self, url, data, **kw):
+        kw.setdefault('headers', {})
+        kw['headers']['Content-Type'] = 'application/json'
+        kw['headers']['Accept'] = 'application/json'
 
-        res = requests.post(
-            url, data=json.dumps({'token': access_token}),
-            headers={'Content-Type': 'application/json',
-                     'Accept': 'application/json'})
         try:
+            res = requests.post(url, json.dumps(data), **kw)
             res.raise_for_status()
         except HTTPError, exc:
-            log.info(u'FxA access token exception: '
-                     u'{exc.__class__.__name__}: {exc}; token={token}; '
-                     u'url={url}; status={status}'
+            log.info(u'FxA client exception: '
+                     u'{exc.__class__.__name__}: {exc}; '
+                     u'url={url}; status={status}; data={data}'
                      .format(exc=exc, url=res.url, status=res.status_code,
-                             token=access_token))
-            log.debug(u'FxA access token response: {res}; token={token}'
-                      .format(res=res.content, token=access_token))
+                             data=data))
             raise forms.ValidationError('invalid FxA response')
 
-        fxa_token = res.json()
+        return res.json()
+
+    def clean_authorization_code(self):
+        code = self.cleaned_data['authorization_code']
+        if not code:
+            return
+
+        url = urlparse.urljoin(settings.FXA_OAUTH_URL, 'v1/token')
+        log.info(u'getting token for code at {url}; code={code}'
+                 .format(url=url, code=code))
+
+        result = self.fxa_post(url, {
+            'code': code,
+            'client_id': settings.FXA_CLIENT_ID,
+            'client_secret': settings.FXA_CLIENT_SECRET,
+        })
+        log.info(u'code for token: {}'.format(result))
+
+        self.token_data = self.validate_access_token(result['access_token'])
+        print 'validated token from code', self.token_data
+
+        return code
+
+    def clean_access_token(self):
+        access_token = self.cleaned_data['access_token']
+        if not access_token:
+            return
+
+        self.token_data = self.validate_access_token(access_token)
+        return access_token
+
+    def validate_access_token(self, access_token):
+        url = urlparse.urljoin(settings.FXA_OAUTH_URL, 'v1/verify')
+        log.info(u'verifying access token at {url}; token={token}'
+                 .format(url=url, token=access_token))
```

Comment:
codes are short lived and tokens only last two weeks. But yeah, good idea. I'll truncate the tokens just in case.